flocker/node/agents/ebs.py

```diff
@@ -0,0 +1,198 @@
+# Copyright Hybrid Logic Ltd.  See LICENSE file for details.
+
+"""
+An EBS implementation of the ``IBlockDeviceAPI``.
+"""
+
+from bitmath import Byte, GB
+import time
+from uuid import UUID
+
+from boto import ec2
+from pyrsistent import PRecord, field
+from zope.interface import implementer
+
+from .blockdevice import IBlockDeviceAPI, BlockDeviceVolume, UnknownVolume
+
+DATASET_ID_LABEL = u'flocker-dataset-id'
+METADATA_VERSION_LABEL = u'flocker-metadata-version'
+CLUSTER_ID_LABEL = u'flocker-cluster-id'
+
+
+def ec2_client(region, zone, access_key_id, secret_access_key):
+    """
+    Establish connection to EC2 client.
+
+    :param String region: The name of the EC2 region to connect to.
+    :param String zone: The zone for the EC2 region to connect to.
+    :param String access_key_id: "aws_access_key_id" credential for EC2.
+    :param String secret_access_key: "aws_secret_access_key" EC2 credential.
+
+    :return: An ``_EC2`` giving information about EC2 client connection
+        and EC2 instance zone.
+    """
+    return _EC2(zone=zone,
+                connection=ec2.connect_to_region(
+                    region,
+                    aws_access_key_id=access_key_id,
+                    aws_secret_access_key=secret_access_key))
+
+
+class _EC2(PRecord):
+    """
+    :ivar string zone: The name of the zone for the connection.
+    :ivar EC2Connection connection: Object representing connection to
+        an EC2 instance.
+    """
+    zone = field()
+    connection = field()
+
+
+@implementer(IBlockDeviceAPI)
+class EBSBlockDeviceAPI(object):
+    """
+    An EBS implementation of ``IBlockDeviceAPI`` which creates
+    block devices in an EC2 cluster using Boto APIs.
+    """
+    def __init__(self, ec2_client, cluster_id):
+        """
+        Initialize EBS block device API instance.
+
+        :param ``_EC2`` ec2_client: PRecord storing EC2
+            client connection and EC2 instance zone.
+        :param String cluster_id: UUID of cluster for this
+            API instance.
+        """
+        self.connection = ec2_client.connection
+        self.zone = ec2_client.zone
+        self.cluster_id = cluster_id
+
+    def _blockdevicevolume_from_ebs_volume(self, ebs_volume):
+        """
+        Helper function to convert Volume information from
+        EBS format to Flocker block device format.
+        """
+        return BlockDeviceVolume(
+            blockdevice_id=unicode(ebs_volume.id),
+            size=int(GB(ebs_volume.size).to_Byte().value),
+            host=None,
+            dataset_id=UUID(ebs_volume.tags[DATASET_ID_LABEL])
+        )
+
+    def _wait_for_volume(self, expected_volume,
+                         expected_status=u'available',
+                         time_limit=60):
+        """
+        Helper function to wait for up to 60s for given volume
+        to be in 'available' state.
```

Comment:
EBS does have 'in-use' state, but the two current usecases for _wait_for_volume (with default target state of 'available') are in create_volume path, where we do want to wait for it to be in 'available' state before declaring success. Since we are not multi-threaded at the moment, there is no possibility of the volume changing state to in-use before the current create-volume returns. Please correct me if this is not the case?